// tools/Brush.js
import {
  RESOLUTION
} from 'config/constants'

import Tool from './Tool'

import { get } from 'svelte/store'

import { project, store } from 'store'

import { DrawPixels } from 'commands'

import { getPixelPositions } from 'lib'

import { renderPixels } from 'lib/drawing'

class Brush extends Tool {
  constructor() {
    super()

    this.selecting = false

    this.name = 'brush'

    this.eraseable = true
  }

  draw(evt) {
    const { x, y } = getPixelPositions(evt)

    const proj = get(project)

    let color = proj.currentFrame.getPixel(x, y) === proj.color ? null : proj.color

    let newColor = this.selecting ? this.selectingColor : proj.color

    if (!this.selecting) {
      this.selecting = true

      this.selectingColor = color
    }

    return (this.pixels[x] = this.pixels[x] || {})[y] = newColor
  }

  render(ctx) {
    renderPixels({ ctx, pixels: this.pixels, color: store.get('eraseModifier') && 'rgba(0,0,0,0.3)' })
  }

  mouseDown(evt) {
    this.pixels = {}

    return this.draw(evt)
  }

  mouseMove(evt) {
    const isMouseDown = evt.buttons == 1

    if (!isMouseDown) { return false }

    return this.draw(evt)
  }

  mouseUp(evt) {
    DrawPixels.perform({ frame: get(project).currentFrame, pixels: this.pixels })

    this.reset()
  }

  reset() {
    super.reset()

    this.pixels = {}
    this.selecting = false
  }
}

// Mad instance life
export default new Brush()
