// text/lowercase.js
//

const lowercase = {
  a: {
    0: [2, 3, 4],
    1: [1, 4],
    2: [1, 2, 3, 4]
  },
  b: {
    0: [0, 1, 2, 3, 4],
    1: [1, 4],
    2: [1, 2, 3, 4]
  },
  c: {
    0: [1, 2, 3, 4],
    1: [1, 4],
    2: [1, 4]
  },
  d: {
    0: [1, 2, 3, 4],
    1: [1, 4],
    2: [0, 1, 2, 3, 4]
  },
  e: {
    0: [1, 2, 3, 4],
    1: [1, 4],
    2: [1, 2, 4]
  },
  f: {
    0: [2],
    1: [0, 1, 2, 3, 4],
    2: [0, 2]
  },
  g: {
    0: [2, 3, 4, 6],
    1: [1, 4, 6],
    2: [1, 2, 3, 4, 5, 6]
  },
  h: {
    0: [0, 1, 2, 3, 4],
    1: [1],
    2: [1, 2, 3, 4]
  },
  i: {
    0: [-1, 1, 2, 3, 4]
  },
  j: {
    0: [5],
    1: [-1, 1,2, 3, 4, 5]
  },
  k: {
    0: [0, 1, 2, 3, 4],
    1: [1, 3],
    2: [2, 4]
  },
  l: {
    0: [0, 1, 2, 3, 4],
    1: [4]
  },
  m: {
    0: [2, 3, 4],
    1: [1],
    2: [2],
    3: [1],
    4: [2, 3, 4]
  },
  n: {
    0: [1, 2, 3, 4],
    1: [1],
    2: [2, 3, 4]
  },
  o: {
    0: [2, 3],
    1: [1, 4],
    2: [2, 3]
  },
  p: {
    0: [1, 2, 3, 4, 5],
    1: [1, 3],
    2: [1, 2, 3]
  },
  q: {
    0: [1, 2, 3],
    1: [1, 3],
    2: [1, 2, 3, 4, 5]
  },
  r: {
    0: [1, 2, 3, 4],
    1: [1],
  },
  s: {
    0: [1, 2, 4],
    1: [1, 3, 4]
  },
  t: {
    0: [1],
    1: [0, 1, 2, 3, 4],
    2: [1, 4]
  },
  u: {
    0: [1, 2, 3, 4],
    1: [4],
    2: [1, 2, 3, 4]
  },
  v: {
    0: [1, 2, 3],
    1: [4],
    2: [1, 2, 3]
  },
  w: {
    0: [1, 2, 3],
    1: [4],
    2: [3],
    3: [4],
    4: [1, 2, 3]
  },
  x: {
    0: [1, 4],
    1: [2, 3],
    2: [1, 4]
  },
  y: {
    0: [1, 2, 3, 4, 6],
    1: [4, 6],
    2: [1, 2, 3, 4, 5, 6]
  },
  z: {
    0: [1, 3, 4],
    1: [1, 2, 4]
  }
}

export default lowercase
