// text/uppercase.js
//

const uppercase = {
  A: {
    0: [1, 2, 3, 4],
    1: [0, 2],
    2: [1, 2, 3, 4]
  },
  B: {
    0: [0, 1, 2, 3, 4],
    1: [0, 2, 4],
    2: [1, 3]
  },
  C: {
    0: [0, 1, 2, 3, 4],
    1: [0, 4],
    2: [0, 4]
  },
  D: {
    0: [0, 1, 2, 3, 4],
    1: [0, 4],
    2: [1, 2, 3]
  },
  E: {
    0: [0, 1, 2, 3, 4],
    1: [0, 2, 4],
    2: [0, 4]
  },
  F: {
    0: [0, 1, 2, 3, 4],
    1: [0, 2],
    2: [0]
  },
  G: {
    0: [0, 1, 2, 3, 4],
    1: [0, 4],
    2: [0, 2, 4],
    3: [0, 2, 3, 4],
  },
  H: {
    0: [0, 1, 2, 3, 4],
    1: [2],
    2: [0, 1, 2, 3, 4]
  },
  I: {
    0: [0, 4],
    1: [0, 1, 2, 3, 4],
    2: [0, 4]
  },
  J: {
    0: [0, 3, 4],
    1: [0, 4],
    2: [0, 1, 2, 3, 4],
    3: [0],
  },
  K: {
    0: [0, 1, 2, 3, 4],
    1: [2],
    2: [1, 3],
    3: [0, 4]
  },
  L: {
    0: [0, 1, 2, 3, 4],
    1: [4],
    2: [4]
  },
  M: {
    0: [0, 1, 2, 3, 4],
    1: [1],
    2: [2],
    3: [1],
    4: [0, 1, 2, 3, 4]
  },
  N: {
    0: [0, 1, 2, 3, 4],
    1: [1],
    2: [2],
    3: [3],
    4: [0, 1, 2, 3, 4]
  },
  O: {
    0: [1, 2, 3],
    1: [0, 4],
    2: [1, 2, 3]
  },
  P: {
    0: [0, 1, 2, 3, 4],
    1: [0, 2],
    2: [0, 1, 2]
  },
  Q: {
    0: [0, 1, 2, 3, 4],
    1: [0, 4],
    2: [0, 3],
    3: [0, 1, 2, 4]
  },
  R: {
    0: [0, 1, 2, 3, 4],
    1: [0, 2],
    2: [0, 1, 3, 4]
  },
  S: {
    0: [0, 1, 4],
    1: [0, 2, 4],
    2: [0, 3, 4]
  },
  T: {
    0: [0],
    1: [0, 1, 2, 3, 4],
    2: [0]
  },
  U: {
    0: [0, 1, 2, 3, 4],
    1: [4],
    2: [0, 1, 2, 3, 4]
  },
  V: {
    0: [0, 1, 2, 3],
    1: [4],
    2: [0, 1, 2, 3]
  },
  W: {
    0: [0, 1, 2, 3],
    1: [4],
    2: [3],
    3: [4],
    4: [0, 1, 2, 3]
  },
  X: {
    0: [0, 1, 3, 4],
    1: [2],
    2: [0, 1, 3, 4]
  },
  Y: {
    0: [0, 1, 2],
    1: [2, 3, 4],
    2: [0, 1, 2]
  },
  Z: {
    0: [0, 3, 4],
    1: [0, 2, 4],
    2: [0, 1, 4]
  }
}

export default uppercase
