// models/Shortcut.js
import { get } from 'svelte/store'

import KEYCODES from 'config/keycodes'

import { PIXEL_SIZE } from 'config/constants'

import { Layer } from 'models'

import { project, store, clipboard, mouse } from 'store'

/*
  Layers
*/

const selectLayer = (number, append = false) => evt => {
  let frame = get(project).currentFrame

  let layerId = frame.layerIds[number - 1]

  if (layerId) {
    let activeLayerIds = [...get(store).activeLayerIds]

    if (append) {
      let index = activeLayerIds.indexOf(layerId)

      if (index !== -1) {
        activeLayerIds.splice(index, 1)
      }
      else {
        activeLayerIds.push(layerId)
      }

      get(project).setTool('pointer')
    }
    else {
      activeLayerIds = [layerId]
    }

    // $project.setTool('pointer')

    store.update({ activeLayerIds })
  }

  return null
}

const layerSelections = () => {
  let commands = {
    0: selectLayer(10),
    cmd_0: selectLayer(10, true)
  }

  for (let i = 1; i < 10; i++) {
    commands[i] = selectLayer(i)
    commands['cmd_' + i] = selectLayer(i, true)
  }

  return commands
}

const joinLayers = evt => {
  let activeLayersIds = get(store).activeLayerIds

  // Need 2 or more layers to merge!
  if (activeLayerIds > 1) {
    let frame = get(project).currentFrame

    let baseLayer = frame.layers[activeLayerIds[0]]

    baseLayer.merge(frame.layers[activeLayerIds[1]])
  }
}

const moveLayer = (x, y) => ({
  conditions: () => document.activeElement.tagName !== 'INPUT',
  method: evt => {
    let multiplier = evt.shiftKey ? 5 : 1

    let frame = get(project).currentFrame,
          ids = get(store).activeLayerIds

    for (let id of ids) {
      frame.layers[id]._x += x * PIXEL_SIZE * multiplier
      frame.layers[id]._y += y * PIXEL_SIZE * multiplier
    }

    return true
  }
})

const changeLayerZIndex = direction => evt => {
  return get(project).currentFrame.changeZIndices(get(store).activeLayerIds, direction)
}

const selectAllLayers = {
  conditions: () => document.activeElement.tagName !== 'INPUT',
  method: evt => {
    get(project).setTool('pointer')

    return store.update({ activeLayerIds: get(project).currentFrame.layerIds })
  }
}

const deleteActiveLayers = evt => {
  get(project).currentFrame.destroyLayers(get(store).activeLayerIds)

  store.update({ activeLayerIds: [] })
}

/* End layers */

const setTool = name => evt => get(project).setTool(name)

/*
  Frames
*/

const FRAMES_IN_ROW = 10 // poor constant placement, FYI.

const switchFrame = (x, y) => evt => {
  const proj = get(project)

  // Move left & right
  if (x !== 0) {
    proj.frameIndex = Math.max(0, Math.min(proj.frameIndex + x, proj.frames.length - 1))
  }

  // Move up & down
  if ((y === -1 && proj.frameIndex >= FRAMES_IN_ROW) ||
      (y === +1 && proj.frameIndex < proj.frames.length - FRAMES_IN_ROW))
  {
     proj.frameIndex = Math.max(0, Math.min(proj.frameIndex + y * FRAMES_IN_ROW, proj.frames.length - 1))
  }
}

const addFrame = duplicate => evt => {
  let addBefore = evt.shiftKey

  get(project).addFrame(addBefore, duplicate)
}

const duplicateFrame = addFrame(true)

const deleteFrame = evt => get(project).deleteCurrentFrame()

/* End frames */


/*
  Cut, copy & paste
*/

const _copy = (callback = null) => evt => {
  const frame = get(project).currentFrame

  const activeLayers = get(store).activeLayerIds.map(id => frame.layers[id])

  callback && callback(evt)

  clipboard.set(activeLayers)
}

const copy = _copy()

const cut = _copy(deleteActiveLayers)

const combinedBounds = layers => {
  return layers.reduce((coords, layer) => {
    coords[0] = Math.min(coords[0], layer._x + layer.bounds[0])
    coords[1] = Math.min(coords[1], layer._y + layer.bounds[1])
    coords[2] = Math.max(coords[2], layer._x + layer.bounds[2])
    coords[3] = Math.max(coords[3], layer._y + layer.bounds[3])

    return coords
  }, [Infinity, Infinity, 0, 0])
}

const paste = () => {
  const frame = get(project).currentFrame

  let layers = get(clipboard).map(Layer.duplicate)

  const { x, y } = get(mouse)

  const [xMin, yMin, xMax, yMax] = combinedBounds(layers)

  const hWidth = (xMax - xMin) / 2,
       hHeight = (yMax - yMin) / 2

  for (let layer of layers) {
    layer.id = frame.layerId++

    // Set layer to center of mouse
    layer._x += x - xMin - hWidth
    layer._y += y - yMin - hHeight

    frame.layers[layer.id] = layer

    frame.layerIds = [layer.id, ...frame.layerIds]
  }

  frame.saveImage()

  store.update({ activeLayerIds: layers.map(layer => layer.id) })
}

/* End cut, copy & paste */

/*
  Project controls
*/

const toggleUnderlay = () => {
  let proj = get(project)

  return proj.showUnderlay = !proj.showUnderlay
}

const toggleGrid = () => {
  let proj = get(project)

  return proj.showGrid = !proj.showGrid
}

const play = () => {
  // HMM
}

/* End project controls */

const _escape = evt => {
  const { buttons } = get(mouse)

  if (buttons) {
    get(project).tool.reset()
  }
  else if (get(store).activeLayerIds.length === 0) {
    get(project).setTool('pointer')
  }
  else {
    store.update({ activeLayerIds: [] })
  }
}

/*
  Shortcut pattern = (cmd_)(shift_){ key name }
*/

const SHORTCUTS = {
  // Tools
  p: setTool('pointer'),
  b: setTool('brush'),
  r: setTool('rectangle'),
  o: setTool('ellipse'),
  f: setTool('fill'),
  l: setTool('line'),

  // Layers
  left:  moveLayer(-1, 0),
  right: moveLayer(+1, 0),
  up:    moveLayer(0, -1),
  down:  moveLayer(0, +1),

  cmd_j: joinLayers,

  ...layerSelections(),

  cmd_left_square_bracket: changeLayerZIndex(1),
  cmd_right_square_bracket: changeLayerZIndex(-1),

  // Frames
  cmd_f: addFrame(false),
  cmd_d: duplicateFrame,
  cmd_left:  switchFrame(-1, 0),
  cmd_right: switchFrame(+1, 0),
  cmd_up:    switchFrame(0, -1),
  cmd_down:  switchFrame(0, +1),

  cmd_a: selectAllLayers,

  backspace: deleteActiveLayers,
  cmd_backspace: deleteFrame,

  // Cut, copy & paste
  cmd_x: cut,
  cmd_c: copy,
  cmd_v: paste,

  // Project
  cmd_u: toggleUnderlay,
  cmd_l: toggleGrid,

  spacebar: play,

  // Misc
  escape: _escape,
}

const trueFn = (evt) => evt

class Shortcut {
  static call(evt) {
    let key = KEYCODES[evt.keyCode]

    if (!key) return false

    if (evt.metaKey || evt.ctrlKey) {
      key = 'cmd_' + key
    }

    // if (evt.shiftKey) {
    //   key = 'shift_' + key
    // }

    let shortcut = SHORTCUTS[key]

    const isValid = shortcut && ((shortcut.conditions || trueFn)(evt))

    if (isValid) {
      evt.preventDefault()

      let result = (shortcut.method || shortcut)(evt)

      return project.update(proj => {
        proj.currentFrame.saveImage()

        return proj;
      })
    }

    return false
  }
}

export default Shortcut
